src/App.tsx

```diff
@@ -1,24 +1,197 @@
+    } finally {
+      setTempTodo(null);
+      setLoadingTodoIds([]);
+    }
+  };
+
+  const handleClearCompletedTodos = async () => {
+    const deletedTodos = completedTodos.map(todo => handleDeleteTodo(todo.id));
+
+    try {
+      await Promise.all(deletedTodos);
+    } catch {
+      setError('Unable to clear completed todos');
+    }
+  };
+
+  const handleUpdateTodo = useCallback(
+    async (
+      todoId: number,
+      newTodoData: UpdateTodoArgs,
+    ) => {
+      try {
+        setLoadingTodoIds(prevIds => [...prevIds, todoId]);
+        const updatedTodo = await updateTodo(todoId, newTodoData);
+
+        setTodos(prevTodos => prevTodos.map(todo => {
+          if (todo.id !== todoId) {
+            return todo;
+          }
+
+          return updatedTodo;
+        }));
+      } catch {
+        setError('Unable to update a todo');
+      } finally {
+        setLoadingTodoIds([]);
+      }
+    },
+    [todos],
+  );
+
+  const toggleAllCompletedTodos = async () => {
+    const updatedTodos = todos.map(todo => ({
+      ...todo,
+      completed: !areAllCompleted,
+    }));
+
+    setTodos(updatedTodos);
+
+    try {
+      await Promise.all(
+        updatedTodos.map(todo => (
+          handleUpdateTodo(todo.id, { completed: !areAllCompleted }))),
+      );
+    } catch {
+      setError('Unable to update todos');
+    }
+  };
+
+  const visibleTodos = useMemo(() => {
+    switch (filter) {
+      case FilteringOption.active:
+        return activeTodos;
+
+      case FilteringOption.completed:
+        return completedTodos;
+
+      default:
+        return todos;
+    }
+  }, [filter, todos]);
+
   if (!USER_ID) {
     return <UserWarning />;
   }
 
   return (
-    <section className="section container">
-      <p className="title is-4">
-        Copy all you need from the prev task:
-        <br />
-        <a href="https://github.com/mate-academy/react_todo-app-add-and-delete#react-todo-app-add-and-delete">React Todo App - Add and Delete</a>
-      </p>
-
-      <p className="subtitle">Styles are already copied</p>
-    </section>
+    <div className="todoapp">
+      <h1 className="todoapp__title">todos</h1>
+
+      <div className="todoapp__content">
+        <Header
+          setError={setError}
+          onAdd={handleAddTodo}
+          tempTodo={tempTodo}
+          toggleAllCompletedTodos={toggleAllCompletedTodos}
+          areAllCompleted={areAllCompleted}
+        />
+        <TodoList
+          todos={visibleTodos}
+          onDelete={handleDeleteTodo}
+          loadingTodoIds={loadingTodoIds}
+          tempTodo={tempTodo}
+          handleUpdateTodo={handleUpdateTodo}
+        />
+
+        <Footer
+          todos={visibleTodos}
+          filter={filter}
+          setFilter={setFilter}
+          activeTodos={activeTodos}
+          completedTodos={completedTodos}
+          handleClearCompletedTodos={handleClearCompletedTodos}
+        />
+      </div>
+      <Notifications
```

Comment:
use more descriptive name
```suggestion
      <ErrorNotifications
```